Editorial suggestions from Benjamin Kaduk (transport) (#4597)
quicwg/base-drafts: draft-ietf-quic-transport.md

The values provided by a peer for these transport parameters MUST match the
values that an endpoint used in the Destination and Source Connection ID fields
of Initial packets that it sent (and received, for servers). Endpoints MUST
validate that received transport parameters match received Connection ID values.

> Including connection ID values in transport

Author: I'm calling this "editorial" even though it adds a new MUST, because I believe the intent was already that you do this, and so the text is just reiterating an existing expectation.

require multiple round trips or retransmissions of this data.

The payload of an Initial packet includes a CRYPTO frame (or frames) containing
a cryptographic handshake message, ACK frames, CONNECTION_CLOSE frames of type
0x1c, or some combination thereof.  PING and PADDING frames are also permitted.

> An endpoint that

Author: This is pretty awkward.  The idea is that the old text implies that it is invalid to have an Initial with neither a CRYPTO nor an ACK.  But IIUC it's valid to send CONNECTION_CLOSE and nothing else...

Author: Done.

A Retry packet (shown in {{retry-format}}) does not contain any protected
fields.  The value in the Unused field is set to an arbitrary value by the
server; a client MUST NOT process those bits other than to verify the Retry

> Integrity Tag.  In addition to the fields from the

Author: Okay.  I think we have ample precedent for "process" being a flexible term...

Packet Number:

: The packet number field is 1 to 4 bytes long. The packet number is protected

>   using header protection; see

Author: Trying to use consistent language for each field covered by header protection.

initial_max_streams_bidi (0x08):

: The initial maximum bidirectional streams parameter is an integer value that
  contains the initial maximum number of bidirectional streams the endpoint

>   that receives this transport parameter is

Author: We were inconsistent about how we described the two endpoints in these paragraphs.  I'm not sure I caught all instances, though, but the "the endpoint that receives this transport parameter" seems like a good clear description.

: The disable active migration transport parameter is included if the endpoint
  does not support active connection migration ({{migration}}) on the address
  being used during the handshake.  An endpoint that receives this transport

>   parameter MUST NOT use a a new local address when sending to the address that

Reviewer: ```suggestion
  parameter MUST NOT use a new local address when sending to the address that
```

Receivers send ACK frames (types 0x02 and 0x03) to inform senders of packets
they have received and processed. The ACK frame contains one or more ACK Ranges.
ACK Ranges identify acknowledged packets. If the frame type is 0x03, ACK frames

> also contain the cumulative count of QUIC packets with associated ECN marks received on the

Author: (you can't take a sum of QUIC packets; there is no addition operation defined for them.)

ACK Range Count:

> : A variable-length integer specifying the number of ACK Range fields in

Author: This whole section was a bit awkward.  I tried to operate on the principle that the Ack Range is the compound structure, and that Gap and the Ack Range Length are the member fields within it.  The resulting text isn't very pretty, though...

Reviewer: This is a good improvement -- thanks!

properties. Any attack on the TLS handshake could affect QUIC.

Any attack on the TLS handshake that compromises the secrecy or uniqueness
of session keys, or the authentication of the participating peers, affects other security guarantees provided by QUIC that depend

> on those keys. For instance, migration ({{migration}}) depends on the efficacy

Author: I'm open to better phrasing suggestions here, too.

Reviewer 2: The structure doesn't seem ideal, but this is OK with me and I can't see anything better.

negotiation is folded into the TLS transcript and thus provides the same
integrity guarantees as ordinary TLS negotiation.  An attacker can observe
the client's transport parameters (as long as it knows the version-specific
salt) but cannot observe the server's transport parameters as sent to
the client and cannot influence
parameter negotiation.  (The attacker could make its own connection to
the server and observe the transport parameters sent on that

> connection.)

Reviewer 2: I don't think that this adds much value and so I would prefer not to take this change.

Author: Okay; reverted.

Connection IDs are unencrypted but integrity protected in all packets.

> This version of QUIC does not incorporate a complete or secure version negotiation mechanism;

Reviewer 2: This is unnecessary.

Author: Okay; reverted.

key material established during the handshake; see {{handshake}} and
{{handshake-properties}}. Similarly, any active attacker that observes packets
and attempts to insert new data or modify existing data in those packets should

> not be able to generate non-Initial packets deemed valid by the receiving endpoint.

Reviewer 2: ```suggestion
not be able to generate packets deemed valid by the receiving endpoint,
other than Initial packets.
```

- Delay packets
- Reorder packets
- Drop packets

> - Split and merge datagrams along packet boundaries (though header protection obfsucates the location of packet boundaries)

Reviewer 2: Unnecessary change.

Author: Reverted

lifetime of address validation tokens; see {{validate-future}}.

> ## Optimistic ACK Attack (#optimistic-ack-attack}

Author: ```suggestion
## Optimistic ACK Attack {#optimistic-ack-attack}
```

lifetime of address validation tokens; see {{validate-future}}.

## Optimistic ACK Attack (#optimistic-ack-attack}

An endpoint that acknowledges packets it has not received might cause a
congestion controller to permit sending at rates beyond what the network
supports.  An endpoint MAY skip packet numbers when sending packets to detect
this behavior.  An endpoint can then immediately close the connection with a
connection error of type PROTOCOL_VIOLATION; see {{immediate-close}}.
Note that introducing artificial gaps might increase the size of the

> returned ACK packets in a manner that scales roughly llinearly with the

Author: Oops, "llinearly" typo.
```suggestion
returned ACK packets in a manner that scales roughly linearly with the
```

> number of gaps introduced, so fewer, larger, gaps may be more efficient.

Reviewer 2: Unnecessary change.

Author: Reverted.

The initial contents of this registry are shown in {{iana-error-table}}.

> | Value | Code                      | Description                   | Specification   |

Author: (The prose says this field is 'Code', not 'Error'.)
